File: codex-rs/core/src/tools/sandboxing.rs
//! Shared approvals and sandboxing traits used by tool runtimes.
//!
//! Consolidates the approval flow primitives (`ApprovalDecision`, `ApprovalStore`,
//! `ApprovalCtx`, `Approvable`) together with the sandbox orchestration traits
//! and helpers (`Sandboxable`, `ToolRuntime`, `SandboxAttempt`, etc.).

use crate::codex::Session;
use crate::error::CodexErr;
use crate::sandboxing::CommandSpec;
use crate::sandboxing::SandboxManager;
use crate::sandboxing::SandboxTransformError;
use codex_protocol::protocol::AskForApproval;
use codex_protocol::protocol::ReviewDecision;
use serde::Serialize;
use std::collections::HashMap;
use std::fmt::Debug;
use std::hash::Hash;
use std::path::Path;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ApprovalDecision {
    Approved,
    ApprovedForSession,
    Denied,
    Abort,
}

impl From<ReviewDecision> for ApprovalDecision {
    fn from(value: ReviewDecision) -> Self {
        match value {
            ReviewDecision::Approved => ApprovalDecision::Approved,
            ReviewDecision::ApprovedForSession => ApprovalDecision::ApprovedForSession,
            ReviewDecision::Denied => ApprovalDecision::Denied,
            ReviewDecision::Abort => ApprovalDecision::Abort,
        }
    }
}

#[derive(Clone, Default, Debug)]
pub(crate) struct ApprovalStore {
    // Store serialized keys for generic caching across requests.
    map: HashMap<String, ApprovalDecision>,
}

impl ApprovalStore {
    pub fn get<K>(&self, key: &K) -> Option<ApprovalDecision>
    where
        K: Serialize,
    {
        let s = serde_json::to_string(key).ok()?;
        self.map.get(&s).cloned()
    }

    pub fn put<K>(&mut self, key: K, value: ApprovalDecision)
    where
        K: Serialize,
    {
        if let Ok(s) = serde_json::to_string(&key) {
            self.map.insert(s, value);
        }
    }
}

#[derive(Clone)]
pub(crate) struct ApprovalCtx<'a> {
    pub session: &'a Session,
    pub sub_id: &'a str,
    pub call_id: &'a str,
    pub retry_reason: Option<String>,
}

pub(crate) trait Approvable<Req> {
    type ApprovalKey: Hash + Eq + Clone + Debug + Serialize;

    fn approval_key(&self, req: &Req) -> Self::ApprovalKey;

    fn should_bypass_approval(&self, policy: AskForApproval) -> bool {
        matches!(policy, AskForApproval::Never)
    }

    fn start_approval_async<'a>(
        &'a mut self,
        req: &'a Req,
        ctx: ApprovalCtx<'a>,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = ApprovalDecision> + Send + 'a>>;
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum SandboxablePreference {
    Auto,
    #[allow(dead_code)] // Will be used by later tools.
    Require,
    #[allow(dead_code)] // Will be used by later tools.
    Forbid,
}

pub(crate) trait Sandboxable {
    fn sandbox_preference(&self) -> SandboxablePreference;
    fn escalate_on_failure(&self) -> bool {
        true
    }
}

pub(crate) struct ToolCtx<'a> {
    pub session: &'a Session,
    pub sub_id: String,
    pub call_id: String,
}

#[derive(Debug)]
pub(crate) enum ToolError {
    Rejected(String),
    SandboxDenied(String),
    Codex(CodexErr),
}

pub(crate) trait ToolRuntime<Req, Out>: Approvable<Req> + Sandboxable {
    async fn run(
        &mut self,
        req: &Req,
        attempt: &SandboxAttempt<'_>,
        ctx: &ToolCtx,
    ) -> Result<Out, ToolError>;
}

pub(crate) struct SandboxAttempt<'a> {
    pub sandbox: crate::exec::SandboxType,
    pub policy: &'a crate::protocol::SandboxPolicy,
    pub(crate) manager: &'a SandboxManager,
    pub(crate) sandbox_cwd: &'a Path,
    pub codex_linux_sandbox_exe: Option<&'a std::path::PathBuf>,
}

impl<'a> SandboxAttempt<'a> {
    pub fn env_for(
        &self,
        spec: &CommandSpec,
    ) -> Result<crate::sandboxing::ExecEnv, SandboxTransformError> {
        self.manager.transform(
            spec,
            self.policy,
            self.sandbox,
            self.sandbox_cwd,
            self.codex_linux_sandbox_exe,
        )
    }
}

File: src/nativeMain/kotlin/ai/solace/coder/core/tools/Sandboxing.kt
// port-lint: source core/src/tools/sandboxing.rs
package ai.solace.coder.core.tools

import ai.solace.coder.core.error.CodexError
import ai.solace.coder.core.session.Session
import ai.solace.coder.core.session.TurnContext
import ai.solace.coder.exec.sandbox.CommandSpec
import ai.solace.coder.exec.sandbox.ExecEnv
import ai.solace.coder.exec.sandbox.SandboxManager
import ai.solace.coder.exec.sandbox.SandboxTransformError
import ai.solace.coder.exec.sandbox.SandboxType
import ai.solace.coder.protocol.AskForApproval
import ai.solace.coder.protocol.ReviewDecision
import ai.solace.coder.protocol.SandboxCommandAssessment
import ai.solace.coder.protocol.SandboxPolicy
import kotlinx.coroutines.sync.Mutex
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class ApprovalStore {
    private val mutex = Mutex()
    private val map = mutableMapOf<String, ReviewDecision>()

    suspend fun <K> get(key: K): ReviewDecision? where K : Any {
        return mutex.withLock {
            val s = key.toString()
            map[s]
        }
    }

    suspend fun <K> put(key: K, value: ReviewDecision) where K : Any {
        mutex.withLock {
            val s = key.toString()
            map[s] = value
        }
    }
}

suspend fun <K> withCachedApproval(
    services: SessionServices,
    key: K,
    fetch: suspend () -> ReviewDecision
): ReviewDecision where K : Any {
    val store = services.toolApprovals
    
    store.get(key)?.let { return it }
    
    val decision = fetch()
    
    if (decision == ReviewDecision.ApprovedForSession) {
        store.put(key, decision)
    }
    
    return decision
}

// Placeholder for SessionServices
class SessionServices {
    val toolApprovals = ApprovalStore() // Should be Mutex protected
}

data class ApprovalCtx(
    val session: Session,
    val turn: TurnContext,
    val callId: String,
    val retryReason: String?,
    val risk: SandboxCommandAssessment?
)

sealed class ApprovalRequirement {
    data class Skip(val bypassSandbox: Boolean) : ApprovalRequirement()
    data class NeedsApproval(val reason: String?) : ApprovalRequirement()
    data class Forbidden(val reason: String) : ApprovalRequirement()
}

fun defaultApprovalRequirement(
    policy: AskForApproval,
    sandboxPolicy: SandboxPolicy
): ApprovalRequirement {
    val needsApproval = when (policy) {
        AskForApproval.Never, AskForApproval.OnFailure -> false
        AskForApproval.OnRequest -> sandboxPolicy != SandboxPolicy.DangerFullAccess
        AskForApproval.UnlessTrusted -> true
    }

    return if (needsApproval) {
        ApprovalRequirement.NeedsApproval(null)
    } else {
        ApprovalRequirement.Skip(bypassSandbox = false)
    }
}

enum class SandboxOverride {
    NoOverride,
    BypassSandboxFirstAttempt
}

interface Approvable<Req> {
    // type ApprovalKey
    // fun approvalKey(req: Req): ApprovalKey

    fun sandboxModeForFirstAttempt(req: Req): SandboxOverride {
        return SandboxOverride.NoOverride
    }

    fun shouldBypassApproval(policy: AskForApproval, alreadyApproved: Boolean): Boolean {
        if (alreadyApproved) {
            return true
        }
        return policy == AskForApproval.Never
    }

    fun approvalRequirement(req: Req): ApprovalRequirement? {
        return null
    }

    fun wantsNoSandboxApproval(policy: AskForApproval): Boolean {
        return policy != AskForApproval.Never && policy != AskForApproval.OnRequest
    }

    suspend fun startApprovalAsync(req: Req, ctx: ApprovalCtx): ReviewDecision
}

enum class SandboxablePreference {
    Auto,
    Require,
    Forbid
}

interface Sandboxable {
    fun sandboxPreference(): SandboxablePreference
    fun escalateOnFailure(): Boolean {
        return true
    }
}

data class ToolCtx(
    val session: Session,
    val turn: TurnContext,
    val callId: String,
    val toolName: String
)

data class SandboxRetryData(
    val command: List<String>,
    val cwd: String // PathBuf -> String
)

interface ProvidesSandboxRetryData {
    fun sandboxRetryData(): SandboxRetryData?
}

sealed class ToolError {
    data class Rejected(val reason: String) : ToolError()
    data class Codex(val error: CodexError) : ToolError()
}

interface ToolRuntime<Req, Out> : Approvable<Req>, Sandboxable {
    suspend fun run(
        req: Req,
        attempt: SandboxAttempt,
        ctx: ToolCtx
    ): Result<Out> // Using Result<Out> which can wrap ToolError logic or throw
}

class SandboxAttempt(
    val sandbox: SandboxType,
    val policy: SandboxPolicy,
    val manager: SandboxManager,
    val sandboxCwd: String, // Path -> String
    val codexLinuxSandboxExe: String? // PathBuf -> String
) {
    fun envFor(spec: CommandSpec): Result<ExecEnv> {
        return manager.transform(
            spec,
            policy,
            sandbox,
            sandboxCwd,
            codexLinuxSandboxExe
        )
    }
}
